### PR TITLE
[SPARK-30735][SQL] New functionality for partitionBy to enhance performance when writing data with partitionBy

### PR DESCRIPTION
**Pull request by tomtongue**
### What changes were proposed in this pull request?

Adding new boolean parameter to `partitionBy` method in DataFrameWriter.

```scala
  @scala.annotation.varargs
  def partitionBy(needRepartition: Boolean, colNames: String*): DataFrameWriter[T] = {
    this.partitioningColumns = Option(colNames)
    if(needRepartition) needRepartitionBasedColumns = true
    this
  }
```

If `partitionBy(true, cols)` is specified, the `needRepartitionBasedColumns` parameter sets `true`. Based on this parameter, when writing data with `saveToV1Source()`, the `repartition` method based on columns is called before writing data.

```scala
  private def saveToV1Source(): Unit = {
    partitioningColumns.foreach { columns =>
      extraOptions += (DataSourceUtils.PARTITIONING_COLUMNS_KEY ->
        DataSourceUtils.encodePartitioningColumns(columns))
    }

    // Code path for data source v1.
    var dfx = df
    if(needRepartitionBasedColumns) {
      dfx = df.repartition(partitioningColumns.toArray.flatten.map(df.col): _*)
    }

    runCommand(dfx.sparkSession, "save") {
      DataSource(
        sparkSession = dfx.sparkSession,
        className = source,
        partitionColumns = partitioningColumns.getOrElse(Nil),
        options = extraOptions.toMap).planForWriting(mode, dfx.logicalPlan)
    }
  }
```

### Why are the changes needed?

This new functionality is possible to help many Spark users who use partitionBy when writing data with DataFrame.

### Does this PR introduce any user-facing change?

Basically not. This new functionality enables user to specify additional boolean parameter for `partitionBy`. When specifying this parameter such as `partitionby(true, cols)`, the repartition method based on specified columns is internally called and is expected to enhance performance and to decrease memory usage when writing.

### How was this patch tested?

Not yet. However, I confirmed that it works by actual testing with DataFrameReader and DataFrameWriter

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Why don't we just explicitly repartition? it doesn't look particularly worth adding another method with the parameter.
